File: python/euler41.py
# Какое существует наибольшее n-значное пан-цифровое простое число?

from datetime import datetime
from functools import lru_cache
from itertools import permutations


@lru_cache(2 ** 5)
def isprime(number: int) -> bool:
    """
    Проверка числа на простоту перебором делителей.

    Args:
        number: Число для проверки.

    Returns:
        True если число простое, иначе False.
    """
    if number in {2, 3, 5, 7}:
        return True
    if number < 2 or number % 2 == 0:
        return False
    if number % 3 == 0 or number % 5 == 0:
        return False
    a, k = number, 0
    for i in range(5, int(number ** 0.5), 6):
        k += 1 if a % i == 0 or a % (i + 2) == 0 else 0
    return k <= 0


def euler41() -> None:
    """
    Решение задачи Эйлера №41.

    Какое существует наибольшее n-значное пан-цифровое простое число?
    """
    n = 9

    # 1 вариант
    start_time = datetime.now()
    pan_primes = []
    for z in range(n, 1, -1):
        pan_primes_check = [
            int(''.join(str(_) for _ in x))
            for x in permutations(range(1, z + 1), z)
            if isprime(int(''.join(str(_) for _ in x)))
        ]
        if pan_primes_check:
            pan_primes.append(max(pan_primes_check))
    print(max(pan_primes))
    print(datetime.now() - start_time)

    # 2 вариант
    start_time = datetime.now()
    pass
    print(datetime.now() - start_time)


if __name__ == "__main__":
    euler41()

File: go/euler41.go
// euler41 находит наибольшее n-значное пан-цифровое простое число.
//
// Решение задачи Эйлера №41.
package main

import (
	"fmt"
	"math"
	"time"
)

func isPrime(number int) bool {
	if number < 2 {
		return false
	}
	if number == 2 || number == 3 || number == 5 || number == 7 {
		return true
	}
	if number%2 == 0 || number%3 == 0 || number%5 == 0 {
		return false
	}
	sqrt := int(math.Sqrt(float64(number)))
	for i := 5; i <= sqrt; i += 6 {
		if number%i == 0 || number%(i+2) == 0 {
			return false
		}
	}
	return true
}

func euler41() {
	n := 9

	startTime := time.Now()
	maxPanPrime := 0
	for z := n; z > 1; z-- {
		// Generate all permutations of 1..z
		perms := generatePermutations(z)
		for _, p := range perms {
			if p[0] == 0 {
				continue
			}
			num := 0
			for _, d := range p {
				num = num*10 + d
			}
			if isPrime(num) && num > maxPanPrime {
				maxPanPrime = num
			}
		}
		if maxPanPrime > 0 {
			break
		}
	}
	fmt.Println(maxPanPrime)
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	fmt.Println(time.Since(startTime))
}

func generatePermutations(n int) [][]int {
	arr := make([]int, n)
	for i := 0; i < n; i++ {
		arr[i] = i + 1
	}
	return permute(arr)
}

func permute(arr []int) [][]int {
	result := [][]int{}
	permuteHelper(arr, 0, &result)
	return result
}

func permuteHelper(arr []int, i int, result *[][]int) {
	if i == len(arr) {
		tmp := make([]int, len(arr))
		copy(tmp, arr)
		*result = append(*result, tmp)
	} else {
		for j := i; j < len(arr); j++ {
			arr[i], arr[j] = arr[j], arr[i]
			permuteHelper(arr, i+1, result)
			arr[i], arr[j] = arr[j], arr[i]
		}
	}
}

func main() {
	euler41()
}
